community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/DefaultRelationshipGroupCursor.java

```diff
@@ -124,30 +138,93 @@ public boolean next()
     {
         if ( isBuffered() )
         {
-            bufferedGroup = bufferedGroup.next;
-            if ( bufferedGroup == null )
-            {
-                return false; // we never have both types of traversal, so terminate early
-            }
-            setType( bufferedGroup.label );
-            setFirstOut( bufferedGroup.outgoing() );
-            setFirstIn( bufferedGroup.incoming() );
-            setFirstLoop( bufferedGroup.loops() );
-            return true;
+            return nextFromBuffer();
+        }
+
+        //We need to check tx state if there are new types added
+        //on the first call of next
+        if ( !hasCheckedTxState )
+        {
+            checkTxStateForUpdates();
+            hasCheckedTxState = true;
         }
 
         do
         {
             if ( getNext() == NO_ID )
             {
-                return false;
+                //We have now run out of groups from the store, however there may still
+                //be new types that was added in the transaction that we haven't visited yet.
+                return nextFromTxState();
             }
             read.group( this, getNext(), page );
         } while ( !inUse() );
 
+        markTypeAsSeen( type() );
         return true;
     }
 
+    private boolean nextFromBuffer()
+    {
+        bufferedGroup = bufferedGroup.next;
+        if ( bufferedGroup == null )
+        {
+            return false; // we never have both types of traversal, so terminate early
+        }
+        setType( bufferedGroup.label );
+        setFirstOut( bufferedGroup.outgoing() );
+        setFirstIn( bufferedGroup.incoming() );
+        setFirstLoop( bufferedGroup.loops() );
+        return true;
+    }
+
+    private boolean nextFromTxState()
+    {
+        if ( txTypeIterator == null && !txTypes.isEmpty() )
+        {
+            txTypeIterator = txTypes.iterator();
+            //here it may be tempting to do txTypes.clear()
+            //however that will also clear the iterator
+        }
+        if ( txTypeIterator != null && txTypeIterator.hasNext() )
+        {
+            setType( txTypeIterator.next() );
```

Comment:
When reading this I had a spider sense tingling. I wonder if there is some problem with state (the relationship pointers) leaking from a previous in-store group when we switch to tx-state groups. WDYT about adding some existing group to the test case, or making a separate test case where there is both existing and added groups?

Reply:
Which I see you already noticed, tested and fixed in the next commit. Nice.